packages/orchestration/src/orchestration-chat-completion-stream.ts

```diff
@@ -0,0 +1,179 @@
+import { createLogger } from '@sap-cloud-sdk/util';
+import { SseStream } from '@sap-ai-sdk/core';
+import { OrchestrationChatCompletionStreamChunkResponse } from './orchestration-chat-completion-stream-chunk-response.js';
+import type {
+  CompletionPostResponseStreaming,
+  LLMChoiceStreaming
+} from './client/api/schema/index.js';
+import type { HttpResponse } from '@sap-cloud-sdk/http-client';
+import type { OrchestrationChatCompletionStreamResponse } from './orchestration-chat-completion-stream-response.js';
+
+const logger = createLogger({
+  package: 'orchestration',
+  messageContext: 'orchestration-chat-completion-stream'
+});
+
+/**
+ * Chat completion stream containing post-processing functions.
+ */
+export class OrchestrationChatCompletionStream<Item> extends SseStream<Item> {
+  /**
+   * Create a chat completion stream based on the http response.
+   * @param response - Http response.
+   * @returns Chat completion stream.
+   * @internal
+   */
+  public static _create(
+    response: HttpResponse,
+    controller: AbortController
+  ): OrchestrationChatCompletionStream<CompletionPostResponseStreaming> {
+    const stream =
+      SseStream.transformToSseStream<CompletionPostResponseStreaming>(
+        response,
+        controller
+      );
+    return new OrchestrationChatCompletionStream(stream.iterator, controller);
+  }
+
+  /**
+   * Wrap raw chunk data with chunk response class to provide helper functions.
+   * @param stream - Chat completion stream.
+   * @internal
+   */
+  static async *_processChunk(
+    stream: OrchestrationChatCompletionStream<CompletionPostResponseStreaming>
+  ): AsyncGenerator<OrchestrationChatCompletionStreamChunkResponse> {
+    for await (const chunk of stream) {
+      yield new OrchestrationChatCompletionStreamChunkResponse(chunk);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  static async *_processFinishReason(
+    stream: OrchestrationChatCompletionStream<OrchestrationChatCompletionStreamChunkResponse>,
+    response?: OrchestrationChatCompletionStreamResponse<OrchestrationChatCompletionStreamChunkResponse>
+  ): AsyncGenerator<OrchestrationChatCompletionStreamChunkResponse> {
+    for await (const chunk of stream) {
+      chunk.data.orchestration_result?.choices.forEach(
+        (choice: LLMChoiceStreaming) => {
+          const choiceIndex = choice.index;
+          if (choiceIndex >= 0) {
+            const finishReason = chunk.getFinishReason(choiceIndex);
+            if (finishReason) {
+              if (response) {
+                response._getFinishReasons().set(choiceIndex, finishReason);
+              }
+              switch (finishReason) {
+                case 'content_filter':
```

Comment:
[q] Any idea, if we can have more finish reasons here other than the ones similar to Open AI?